src/x509.c

```diff
@@ -5882,6 +5882,109 @@ int wolfSSL_X509_cmp(const WOLFSSL_X509 *a, const WOLFSSL_X509 *b)
+                "Digital Signature",
+                "Non Repudiation",
+                "Key Encipherment",
+                "Data Encipherment",
+                "Key Agreement",
+                "Certificate Sign",
+                "CRL Sign",
+                "Encipher Only",
+                "Decipher Only"
+            };
+
+            if (wolfSSL_BIO_write(bio, "            X509v3 Key Usage: \n",
+                    (int)XSTRLEN("            X509v3 Key Usage: \n")) <= 0) {
+                return WOLFSSL_FAILURE;
+            }
+
+            XSNPRINTF(tmp, sizeof(tmp), "                ");
+            for (i = 0; i < sizeof(usages) / sizeof(usages[i]); i++) {
+                if (x509->keyUsage & usages[i]) {
+                    XMEMSET(val, 0, sizeof(val));
+                    ++usageCount;
+                    if (usageCount > 1) {
+                        XSNPRINTF(val, sizeof(val), ", ");
+                    }
+                    XSTRNCAT(val, usageStrs[i], sizeof(val) - XSTRLEN(val) - 1);
+                    XSTRNCAT(tmp, val, sizeof(tmp) - XSTRLEN(tmp) - 1);
+                }
+            }
+            len = (int)XSTRLEN("\n");
+            XSTRNCAT(tmp, "\n", len + 1);
+            if (wolfSSL_BIO_write(bio, tmp, (int)XSTRLEN(tmp)) <= 0) {
+                return WOLFSSL_FAILURE;
+            }
+        }
+
+        /* print extended key usage */
+        if (x509->extKeyUsageCount > 0 && x509->extKeyUsage != 0) {
+            char tmp[150];
+            word32 i;
+            char val[32];
+            int len = 0;
+            int usageCount = 0;
+            const int usages[] = {
+                EXTKEYUSE_OCSP_SIGN,
+                EXTKEYUSE_TIMESTAMP,
+                EXTKEYUSE_EMAILPROT,
+                EXTKEYUSE_CODESIGN,
+                EXTKEYUSE_CLIENT_AUTH,
+                EXTKEYUSE_SERVER_AUTH
+            };
+            const char* usageStrs[] = {
+                "OCSP Signing",
+                "Time Stamping",
+                "E-mail Protection",
+                "Code Signing",
+                "TLS Web Client Authentication",
+                "TLS Web Server Authentication"
+            };
+
+            if (wolfSSL_BIO_write(bio, "            X509v3 Extended Key Usage: "
+                    "\n", (int)XSTRLEN("            X509v3 Extended Key Usage: "
+                    "\n")) <= 0) {
+                return WOLFSSL_FAILURE;
+            }
+
+            XSNPRINTF(tmp, sizeof(tmp), "                ");
+            for (i = 0; i < sizeof(usages) / sizeof(usages[i]); i++) {
+                if (x509->extKeyUsage & usages[i]) {
+                    XMEMSET(val, 0, sizeof(val));
+                    ++usageCount;
+                    if (usageCount > 1) {
+                        XSNPRINTF(val, sizeof(val), ", ");
+                    }
+                    XSTRNCAT(val, usageStrs[i], sizeof(val) - XSTRLEN(val) - 1);
+                    XSTRNCAT(tmp, val, sizeof(tmp) - XSTRLEN(tmp) - 1);
```

Comment:
I don't think an overrun of `tmp` is handled correctly. Technically `tmp` could be 32 * 6 + "leading spaces 15" + null term 1.

Reply:
See previous comment; I don't see how it's possible for this to overrun as written.